### PR TITLE
    Next hop group with members.

### PR DESCRIPTION
**Pull request by erohsik**
Introduce a new type of Next hop group that allows the application
    to provide the list of next hops that are members of the Next
    hop group along with the list of weights.

### CODE REVIEW

**Comment by itaibaz** [comment]
next hop group member objects are being created this way implicitly, without being returned to the application
we don't support such flow in any place AFAIK
If we go with the approach here, application should then query the next hop list of each group created, then query the next hop id, and thus learn the object ids. This seems like a must step, as application should then be able to manage and remove these next hop group members
So while the motivation for the PR is clear, it creates complexity

**Comment by erohsik** [comment]
> next hop group member objects are being created this way implicitly, without being returned to the application we don't support such flow in any place AFAIK If we go with the approach here, application should then query the next hop list of each group created, then query the next hop id, and thus learn the object ids. This seems like a must step, as application should then be able to manage and remove these next hop group members So while the motivation for the PR is clear, it creates complexity


Thanks for taking a look. An example of the modification workflow is in the doc. With this next hop group, if the next hop group starts with {a,b,c,d} and then transitions to {a,b,e}, then the group is modified with the next hop list set to {a,b,e}. There is no explicit deletion of members c and d or addition of e.

**Comment by itaibaz** [comment]
should we mark then
SAI_NEXT_HOP_GROUP_ATTR_NEXT_HOP_MEMBER_LIST
as not valid for this type of group ?

and add clarification in the MD, that 
1) no next hop group member objects are created here
2) to change next hops set a new full list
3) group can be removed without a need to remove group members first 
?

**Comment by erohsik** [comment]
> should we mark then SAI_NEXT_HOP_GROUP_ATTR_NEXT_HOP_MEMBER_LIST as not valid for this type of group ?
> 
> and add clarification in the MD, that
> 
>     1. no next hop group member objects are created here
> 
>     2. to change next hops set a new full list
> 
>     3. group can be removed without a need to remove group members first
>        ?

Thanks for the review. I have updated the PR.

**Review by kcudnik — APPROVED** [review]

**Comment by rck-innovium** [comment]
@erohsik  This is a useful enhancement. Thanks for that.

I have a comment on the way this is modelled in this proposal. The model is maintaining two different lists: one for the NHs and the other for the weights of the corresponding NH. We cannot atomically add a new NH to a created group- we need \[step-1\] call to add the NH to group, and \[step-2\] another call to set the weight. The weights to be programmed in the h/w are inconsistent between step-1 and step-2.

This is not elegant and tomorrow if we want to support this say for Ordered ECMP, we would need another list to specify the 'Sequence Id'.  We could extend this for every ECMP feature like Fine Grained etc.

A cleaner way would be to do one of the below:

#### Option-1: reuse  sai\_next\_hop\_group\_member:

Make SAI\_NEXT\_HOP\_GROUP\_MEMBER\_ATTR\_NEXT\_HOP\_GROUP\_ID mandatory only if NH Group is NOT of type SAI\_NEXT\_HOP\_GROUP\_TYPE\_ECMP\_WITH\_MEMBERS.

Make the object type of SAI\_NEXT\_HOP\_GROUP\_ATTR\_NEXT\_HOP\_LIST as SAI\_OBJECT\_TYPE\_NEXT\_HOP\_GROUP\_MEMBER and remove SAI\_NEXT\_HOP\_GROUP\_ATTR\_NEXT\_HOP\_MEMBER\_WEIGHT\_LIST

**Option-2: create a new type of sai\_next\_hop\_group\_init\_member**

A challenge with option#1 would be to enforce it via SAI metacheckers and backward compatibility. If option#1 is not possible, then we can consider creating a new object type like below (definitely needs a better name).

```
typedef enum _sai_next_hop_group_init_member_attr_t
{

:
:
    /**
     * @brief Next hop id
     *
     * @type sai_object_id_t
     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
     * @objects SAI_OBJECT_TYPE_NEXT_HOP, SAI_OBJECT_TYPE_NEXT_HOP_GROUP
     */
    SAI_NEXT_HOP_GROUP_INIT_MEMBER_ATTR_NEXT_HOP_ID,
  
    /**
     * @brief Member weights
     *
     *
     * @type sai_uint32_t
     * @flags CREATE_AND_SET
     * @default 1
     */
    SAI_NEXT_HOP_GROUP_INIT_MEMBER_ATTR_WEIGHT,
        
 :
 :
 }
```

**Review by rck-innovium — CHANGES_REQUESTED** [review]
Please consider using a object type instead of maintaining the weights as another list.

**Comment by erohsik** [comment]
> @erohsik  This is a useful enhancement. Thanks for that.
> 
> I have a comment on the way this is modelled in this proposal. The model is maintaining two different lists: one for the NHs and the other for the weights of the corresponding NH. We cannot atomically add a new NH to a created group- we need [step-1] call to add the NH to group, and [step-2] another call to set the weight. The weights to be programmed in the h/w are inconsistent between step-1 and step-2.

Thanks for the review.
The document describes the modify workflow. The application would bulk set the two attributes (new NH list, new weight list) in one step..

**Comment by helloanandhi** [comment]
Creating a generic solution for all next hop group types would be more future extensible.

**Comment by rck-innovium** [comment]
> > @erohsik This is a useful enhancement. Thanks for that.  
> > I have a comment on the way this is modelled in this proposal. The model is maintaining two different lists: one for the NHs and the other for the weights of the corresponding NH. We cannot atomically add a new NH to a created group- we need \[step-1\] call to add the NH to group, and \[step-2\] another call to set the weight. The weights to be programmed in the h/w are inconsistent between step-1 and step-2.
> 
> Thanks for the review. The document describes the modify workflow. The application would bulk set the two attributes (new NH list, new weight list) in one step..

The current solution looks to be a point fix for a specific NHG type like WECPM and not for other NHG type. How would you handle the case of supporting counters per member Nexthop using the single step workflow?

**Comment by kcudnik** [comment]
you will need  to rebase on origin/master and force push to pass metadata check

File: doc/ECMP/NextHopGroup_with_members.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        attr.value.u32list.list = weights.data();
+        group_attr.push_back(attr);
+
+        attr.id = SAI_NEXT_HOP_GROUP_ATTR_NEXT_HOP_LIST;
+        attr.value.objlist.count = (uint32_t)nh_oids.size();
+        attr.value.objlist.list = nh_oids.data();
+        group_attr.push_back(attr);
+
+        sai_status_t status =
+          sai->bulkSet(SAI_OBJECT_TYPE_NEXT_HOP_GROUP,
```
Comment (rck-innovium) [inline doc/ECMP/NextHopGroup_with_members.md:80]: Please update to:

sai\_bulk\_object\_set\_attribute\_fn set\_next\_hop\_groups\_attribute()

Comment (erohsik) [reply doc/ECMP/NextHopGroup_with_members.md]: Thanks for the review. Updated the calls..

**Review by rck-innovium — APPROVED** [review]

File: inc/sainexthopgroup.h [hunk 2]
```diff
@@ -107,6 +110,8 @@ typedef enum _sai_next_hop_group_attr_t
     /**
      * @brief Next hop member list
      *
+     * Not valid when SAI_NEXT_HOP_GROUP_ATTR_TYPE == SAI_NEXT_HOP_GROUP_TYPE_ECMP_WITH_MEMBERS
```
[the remark this answers is not in the document]
Comment (rlhui) [reply inc/sainexthopgroup.h:113]: @erohsik , is this addressed?  good to get @itaibaz  to approve as well.

[the remark this answers is not in the document]
Comment (erohsik) [reply inc/sainexthopgroup.h:113]: @itaibaz Is this what you had in mind?

```
@@ -113,6 +113,7 @@ typedef enum _sai_next_hop_group_attr_t
      * @type sai_object_list_t
      * @flags READ_ONLY
      * @objects SAI_OBJECT_TYPE_NEXT_HOP_GROUP_MEMBER
+     * @validonly SAI_NEXT_HOP_GROUP_ATTR_TYPE == SAI_NEXT_HOP_GROUP_TYPE_DYNAMIC_UNORDERED_ECMP or SAI_NEXT_HOP_GROUP_ATTR_TYPE == SAI_NEXT_HOP_GROUP_TYPE_DYNAMIC_ORDERED_ECMP or SAI_NEXT_HOP_GROUP_ATTR_TYPE == SAI_NEXT_HOP_GROUP_TYPE_FINE_GRAIN_ECMP or SAI_NEXT_HOP_GROUP_ATTR_TYPE == SAI_NEXT_HOP_GROUP_TYPE_PROTECTION or SAI_NEXT_HOP_GROUP_ATTR_TYPE == SAI_NEXT_HOP_GROUP_TYPE_CLASS_BASED
      */
     SAI_NEXT_HOP_GROUP_ATTR_NEXT_HOP_MEMBER_LIST,
```

[the remark this answers is not in the document]
Comment (erohsik) [reply inc/sainexthopgroup.h:113]: I do hit this error though `read only attribute can't be valid only`..

[the remark this answers is not in the document]
Comment (erohsik) [reply inc/sainexthopgroup.h:113]: @itaibaz Based on the above error, can we resolve this comment?

**Comment by tjchadaga** [comment]
@itaibaz, @JaiOCP - could you please help sign off on this?

**Comment by tjchadaga** [comment]
@erohsik - please resolve branch conflicts

**Comment by erohsik** [comment]
> @erohsik - please resolve branch conflicts

Done..

**Review by kcudnik — APPROVED** [review]

**Review by JaiOCP — APPROVED** [review]

**Pull request merged**